pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go

```diff
@@ -32,26 +35,68 @@ type Reconciler struct {
 // +kubebuilder:rbac:groups=core.oam.dev,resources=applicationconfigurations,verbs=get;list;watch;create;update;patch;delete
 func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background()
-	var appdeploy v1alpha2.ApplicationDeployment
-	if err := r.Get(ctx, req.NamespacedName, &appdeploy); err != nil {
+	var appDeploy corev1alpha2.ApplicationDeployment
+	if err := r.Get(ctx, req.NamespacedName, &appDeploy); err != nil {
 		if apierrors.IsNotFound(err) {
-			klog.Info("applicationdeployment is deleted")
+			klog.InfoS("application deployment does not exist", "appDeploy", klog.KRef(req.Namespace, req.Name))
 		}
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
-	klog.InfoS("Start to reconcile ", "application deployment", klog.KObj(&appdeploy))
+	klog.InfoS("Start to reconcile ", "application deployment", klog.KObj(&appDeploy))
 
-	// TODO add reconcile logic here
+	r.handleFinalizer(&appDeploy)
 
+	// Get the target application
+	var targetApp, sourceApp corev1alpha2.Application
+	targetAppName := appDeploy.Spec.TargetApplicationName
+	if err := r.Get(ctx, ktypes.NamespacedName{Namespace: req.Namespace, Name: targetAppName},
+		&targetApp); err != nil {
+		klog.ErrorS(err, "cannot locate target application", "target application",
+			klog.KRef(req.Namespace, targetAppName))
+		return ctrl.Result{}, err
+	}
+	// Get the source application
+	sourceAppName := appDeploy.Spec.SourceApplicationName
+	if sourceAppName == nil {
+		klog.Info("source app fields not filled, we assume it is deployed for the first time")
+	} else if err := r.Get(ctx, ktypes.NamespacedName{Namespace: req.Namespace, Name: *sourceAppName},
+		&sourceApp); err != nil {
+		klog.ErrorS(err, "cannot locate source application", "source application", klog.KRef(req.Namespace,
+			*sourceAppName))
+		return ctrl.Result{}, err
+	}
+	// Get the kubernetes workloads to upgrade from the application
+	targetWorkload, sourceWorkload, err := r.extractWorkload(appDeploy.Spec.ComponentList, &targetApp, &sourceApp)
+	if err != nil {
+		klog.Error(err, "cannot locate the workloads object")
+		return ctrl.Result{}, err
+	}
+	klog.InfoS("get the target workload we need to work on", "targetWorkload", klog.KObj(targetWorkload))
+	if sourceWorkload != nil {
+		klog.InfoS("get the source workload we need to work on", "sourceWorkload", klog.KObj(sourceWorkload))
+	}
+	// TODO: pass these two object to the rollout plan
 	return ctrl.Result{}, nil
 }
 
+func (r *Reconciler) handleFinalizer(appDeploy *corev1alpha2.ApplicationDeployment) {
```

Comment:
what do you prepare to use finalizer for?

Reply:
I plan to make sure that any changes we make to the application is reverted